Serialization: Skip invalid decls during module serialization

tshortli opened the pull request:
When `-enable-lazy-typecheck` is specified, serialization may be expected to run on an AST containing invalid declarations since type checking may happen on-demand, during serialization, in this mode. If the declarations that are invalid are not skipped, then the compiler is likely to crash when attempting to serialize them. Now, invalid declarations are skipped and an error is emitted at the end of serialization to note that serialization failed.

Additionally, a new `-Rmodule-serialization` flag can be specified to request more detailed information about module serialization failures. This would be useful in a situation where lazy typechecking does not produce any diagnostic for some reason, but module serialization fails and more information is therefore required to debug.

Resolves rdar://123260476

tshortli committed:
Serialization: Skip invalid decls during module serialization.

When `-enable-lazy-typecheck` is specified, serialization may be expected to
run on an AST containing invalid declarations since type checking may happen
on-demand, during serialization, in this mode. If the declarations that are
invalid are not skipped, then the compiler is likely to crash when attempting
to serialize them. Now, invalid declarations are skipped and an error is
emitted at the end of serialization to note that serialization failed.

Additionally, a new `-Rmodule-serialization` flag can be specified to request
more detailed information about module serialization failures. This would be
useful in a situation where lazy typechecking does not produce any diagnostic
for some reason, but module serialization fails and more information is
therefore required to debug.

Resolves rdar://123260476

xymus commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -808,6 +808,13 @@ WARNING(sema_import_current_module_with_file,none,
+       (const Decl *))
+ERROR(serialization_failed,none,
+      "serialization of module '%0' failed due to the errors above",
+      (StringRef))
```
You should be able to use `const ModuleDecl*` here similarly to the previous one.

xymus commented:
There's a typo in the PR description on `-Rmodule-serilaization`, it may be worth fixing since we may copy and paste from here.

tshortli commented:
> There's a typo in the PR description on `-Rmodule-serilaization`, it may be worth fixing since we may copy and paste from here.

Thanks, fixed.

tshortli committed:
Serialization: Skip custom attributes with invalid types.

Prevents a crash during module emission with `-enable-lazy-typechecking` when
the type corresponding to a `CustomAttr` cannot be resolved.

tshortli committed:
Serialization: Skip patterns with invalid types.

Prevents a crash during module emission with `-enable-lazy-typechecking` when
the type of a `Pattern` cannot be resolved.

tshortli commented:
@swift-ci please smoke test

tshortli commented:
@swift-ci please smoke test

tshortli commented:
@swift-ci please smoke test Linux
